src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -385,18 +385,18 @@ public void Unload()
 
         internal static void OnProcessExit()
         {
-            lock (s_contextsToUnload)
+            lock (s_allContexts)
             {
                 s_isProcessExiting = true;
-                foreach (var alcAlive in s_contextsToUnload)
+                foreach (var alcAlive in s_allContexts)
                 {
-                    if (alcAlive.Value.TryGetTarget(out AssemblyLoadContext alc))
+                    if (alcAlive.Value.TryGetTarget(out AssemblyLoadContext alc) && alc.IsCollectible)
```

Comment:
Is this going to cause `Unload` event to not be delivered to regular contexts during shutdown? (diff from current behavior)

Reply:
Hmm, you are right, I think we were delivering those events for regular contexts in the past (even before the unloadability work started). That would be a breaking change. It seems we should call the Unload event for all contexts then (but not initiate the actual unloading for the non-collectible contexts)

Reply:
We even have a test to verify this: 
System.Tests.UnloadingAndProcessExitTests.UnloadingEventMustHappenBeforeProcessExitEvent

Reply:
Actually, I've just found that the `InitiateUnload` when called from the `OnProcessExit` just sends the Unload event and sets `_state = InternalState.Unloading`. It doesn't initiate the unloading itself. 
So the problematic thing was actually setting the _state even though we are not unloading the context.